### DISCUSSION TITLE
Can magic initrd perform simple script to expand cpio archive?

### DISCUSSION THREAD

**Discussion by ccy** [ipxe/ipxe#705] [General]
I am using [TP Link TG-3468](https://www.tp-link.com/us/home-networking/pci-adapter/tg-3468/) network adapter.  The network adapter is using `Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 15)`.

I can boot into `ipxe` with the network adapter.

Next, I try to netboot Debian installer with:

```
set inst-dir http://deb.debian.org/debian/dists/stable/main/installer-amd64/current/images/netboot/debian-installer/amd64
kernel ${inst-dir}/linux initrd=initrd.gz auto=true priority=critical
initrd ${inst-dir}/initrd.gz
boot
```

I encounter errors in `initrd`:

```
Aug 12 11:09:29 check-missing-firmware:looking for firmware file rtl_nic/rtl8125b-2.fw requested by r8169
Aug 12 11:09:29 check-missing-firmware:missing firmware files (rtl_nic/rtl8168h-2.fw rtl_niczrtl8125b-2.fw) for r8169
Aug 12 11:09:39 check-missing-firmware:removing and loading kernel module r8169
```

As described in [NetbootFirmware](https://wiki.debian.org/DebianInstaller/NetbootFirmware), I believe the issue is due to the network adapter isn't built with a ROM firmware and Debian cannot include the firmware in `initrd` image may be due to license issue.

There is a workaround solution to add the firmware into `initrd` image using [non-free](https://cdimage.debian.org/cdimage/unofficial/non-free/firmware/stable/current/) firmware:

```
# Download standard netboot initrd.gz
curl -L http://deb.debian.org/debian/dists/stable/main/installer-amd64/current/images/netboot/debian-installer/amd64/initrd.gz > initrd-nonfree.gz

# Append non-free firmware to initrd.gz
curl -L https://cdimage.debian.org/cdimage/unofficial/non-free/firmware/stable/current/firmware.cpio.gz >> initrd-nonfree.gz
```

Network boot into `initrd-nonfree.gz` and `check-missing-firmware` issue is gone.

The above solution need to re-pack the `initrd` with non-free firmware.

The `firmware.cpio.gz` is a cpio archive contains `/firmware/*.deb` files.

I am thinking if a solution to utilize the iPXE **magic initrd** to inject the firmware into standard `initrd`:

```
set inst-dir http://deb.debian.org/debian/dists/stable/main/installer-amd64/current/images/netboot/debian-installer/amd64
kernel ${inst-dir}/linux initrd=initrd.magic auto=true priority=critical
initrd ${inst-dir}/initrd.gz
initrd https://cdimage.debian.org/cdimage/unofficial/non-free/firmware/stable/current/firmware.cpio.gz /firmware.cpio.gz
boot
```

The script still won't work yet as the firmware.cpio.gz isn't expand to `<initrd>/firmware` yet.

**First question**: If we can define a script to expand the cpio archive in magic initrd operation, then we can solve the issue.

**Second question**: Can magic initrd can create non exist directory in initrd.gz?  I can't find the `/firmware/firmware.cpio.gz` in initrd runtime with this script:

    initrd https://cdimage.debian.org/cdimage/unofficial/non-free/firmware/stable/current/firmware.cpio.gz /firmware/firmware.cpio.gz

**Comment by NiKiZe** [comment]
As long as the kernel receives initrds, it will combine them.
If I'm not mistaken paths and filenames are stored as given.
If the directory needs to be created the. There is nothing that prevents you from creating an invited with that directory.

**Comment by ccy** [comment]
Thank you very much.  I didn't realize the solution is that simple:

```
set inst-dir http://deb.debian.org/debian/dists/stable/main/installer-amd64/current/images/netboot/debian-installer/amd64
kernel ${inst-dir}/linux initrd=initrd.magic auto=true priority=critical
initrd ${inst-dir}/initrd.gz
initrd http://cdimage.debian.org/cdimage/unofficial/non-free/firmware/stable/current/firmware.cpio.gz
boot
```
The `*.deb` files in `firmware.cpio.gz` expanded into `/firmware` directory now.
